File: solutions/clojure/say/1/src/say.clj
(ns say)

(def num->str {
    1 "one"
    2 "two"
    3 "three"
    4 "four"
    5 "five"
    6 "six"
    7 "seven"
    8 "eight"
    9 "nine"
    10 "ten"
    11 "eleven"
    12 "twelve"
    13 "thirteen"
    14 "fourteen"
    15 "fifteen"
    16 "sixteen"
    17 "seventeen"
    18 "eighteen"
    19 "nineteen"
    20 "twenty"
    30 "thirty"
    40 "forty"
    50 "fifty"
    60 "sixty"
    70 "seventy"
    80 "eighty"
    90 "ninety"})

(defn number-under-100 [num]
  (if (> num 20)
    (str (num->str (* 10 (quot num 10))) \- (num->str (rem num 10)))
    (num->str num)))

(defn number-under-1000 [num]
  (let [hundred (quot num 100)
        tens (rem num 100)]
    (str (if (> hundred 0) 
           (str 
            (number-under-100 hundred)
            " hundred ")
           "")
         (if (> tens 0)
           (number-under-100 tens)
           ""))))

(defn split-number [num]
  {:billion (quot num 1000000000)
   :million (rem (quot num 1000000) 1000)
   :thousand (rem (quot num 1000) 1000)
   :hundred (rem num 1000)})

(defn number [num]
  (when (or (< num 0)
            (> num 999999999999))
    (throw (IllegalArgumentException. "invalid number")))
  (if (zero? num)
    "zero"
    (let [num (split-number num)
          billion (:billion num)
          million (:million num)
          thousand (:thousand num)
          hundred (:hundred num)]
      (clojure.string/trimr (str 
             (if (> billion 0) 
               (str (number-under-1000 billion) " billion ")
               "")
             (if (> million 0)
               (str (number-under-1000 million) " million ")
               "")
             (if (> thousand 0)
               (str (number-under-1000 thousand) " thousand ")
               "")
             (if (> hundred 0)
               (str (number-under-1000 hundred))
               ""))))))

File: solutions/c/say/1/src/say.c
#include "say.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define MAX_ANSWER_LEN (256)
#define say_over_999(start, name) say(n / start, answer); strcat(*answer, " "); strcat(*answer, name); if (n % start) { strcat(*answer, " "); say(n % start, answer); }


static char *under_20[] = {
  "zero",
  "one",
  "two",
  "three",
  "four",
  "five",
  "six",
  "seven",
  "eight",
  "nine",
  "ten",
  "eleven",
  "twelve",
  "thirteen",
  "fourteen",
  "fifteen",
  "sixteen",
  "seventeen",
  "nineteen"
};

static char *tens[] = {
  "",
  "",
  "twenty",
  "thirty",
  "forty",
  "fifty",
  "sixty",
  "seventy",
  "eighty",
  "ninety",
};

int say(long long n, char **answer) {
  if (n < 0 || n > 999999999999)
    return -1;
  
  if (*answer == NULL)
    *answer = calloc(MAX_ANSWER_LEN, sizeof(char));
  
  if (n < 20)
    strcat(*answer, under_20[n]);
  else if (n >= 20 && n < 100) {
    strcat(*answer, tens[n / 10]);
    if (n % 10) {
      strcat(*answer, "-");
      strcat(*answer, under_20[n % 10]);
    }
  } else if (n >= 100 && n < 1000) {
    strcat(*answer, under_20[n / 100]);
    strcat(*answer, " hundred");
    if (n % 100) {
      strcat(*answer, " ");
      say(n % 100, answer);
    }
  } else if (n >= 1000 && n < 1000000) {
    say_over_999(1000, "thousand");
  } else if (n >= 1000000 && n < 1000000000) {
    say_over_999(1000000, "million");
  } else if (n >= 1000000000 && n < 1000000000000) {
    say_over_999(1000000000, "billion");
  }
  
  return 0;
}

#undef MAX_ANSWER_LEN
#undef say_over_999